ls1intum/Artemis
# `General`: Redesign learning goals and store progress

### Checklist
#### General


#### Server
#### Client

### Motivation and Context



The learning goals feature was rarely used, because it only offered limited benefits to the students and instructor.

### Description


In this pull request, we renamed the _learning goals_ to _competencies_ in the client.
The user interface has been overhauled and now features three circular progress bars to track the progress (% of completed lecture units), confidence (% average score w.r.t the mastery threshold), and mastery (weighted metric of progress and confidence).
A new competency details page allows students to work on the specific course material towards achieving a specific competency.

Exercises and lecture units can now be linked to competencies on the respective management pages (instead of only on the competency edit screen).
The progress for all students in each competency is now persisted to the database in preparation for future work on learning paths.
Updating the progress always happens asynchronously and does not delay the server's response. It is triggered when a new result for an exercise is processed in the participant score scheduler, when the user completes a lecture unit, or when a learning goal/lecture unit is edited by an instructor. The user can also manually update the progress by clicking on the “refresh” button.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Course with some Exercises, a Lecture and Lecture Units

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a Competency
4. Go to any lecture unit or exercise and add it to the competency
5. Submit something to the exercise and grade it, the competency progress should update
6. Complete some lecture units, the competency progress should update

#### Exam Mode Testing



Prerequisites:
- 1 Instructor
- 2 Students
- 1 Exam with all exercise types

1. Log in to Artemis
2. Participate in the exam as a student
3. Make sure that the functionality of all exercises is unchanged

### Review Progress






#### Code Review
#### Manual Tests
#### Exam Mode Test

### [Test Coverage](https://app.codecov.io/gh/ls1intum/Artemis/pull/6050)



| Class/File | Branch | Line |
|------------|-------:|-----:|
| fireworks.component.ts | 100% | 100% |
| learning-goal-rings.component.ts | 100% | 100% |
| learning-goal-selection.component.ts | 84% | 87% |
| learning-goal-card.component.ts | 76% | 93% |
| course-learning-goals-details.component.ts | 76% | 83% |
| course-learning-goals.component.ts | 76% | 97% |
| learningGoal.service.ts | 72% | 92% |
| LearningGoalProgressService.java | 68% | 89% |
| LearningGoalService.java | 100% | 100% |
| LearningGoalResource.java | 80% | 97% |
| CourseService.java | 80% | 92% |

### Screenshots



![course-lgs](https://user-images.githubusercontent.com/102237363/211930786-6e5b117d-1649-4157-843f-a8417386d265.png)

![lg-details](https://user-images.githubusercontent.com/102237363/211930751-c79d503d-dd88-4ff8-bdcc-047866288b68.png)

![lgs-mgmt](https://user-images.githubusercontent.com/102237363/211196021-ad89d1d5-a3b2-432d-8fd3-b18a20a10500.png)
